TASKPROC-244:  Initialize metadata via state data and implement reconcile for k8s

jfongatyelp opened the pull request:
In order to handle tron restarts, we will need to first pass kubernetes_pod_executor a list of KubernetesTaskConfigs from our saved state.

We initialize this in `kubernetes_pod_executor` upon init, filling `task_metadata` with UNKNOWN KubernetesTaskStates for each pod name to speed up initalization.

Reconciliation with said pods will then happen in `reconcile()`, where we will check the state of the pod we are attempting to reconcile with the Kubernetes API, then re-use the exact same logic to set TaskState (etc) as when we handle `PodEvents` based on pod phases.   This required splitting out said logic into two new helpers, `__update_deleted_pod` and `__update_modified_pod` so they could be used for both usecases.

nemacysts commented:
@jfongatyelp 
> 1. If we cannot find the pod after restart, then we treat it as a `killed` pod (i.e. manual operator intervention.).  Should this instead be treated as `lost`?

yea, that's what I was thinking (treating it as `lost`) - this *should* match what we do in Mesos with `unknown` (we just took over `unknown` for a different purpose in k8s :p)

jfongatyelp commented on task_processing/plugins/kubernetes/kubernetes_pod_executor.py:
```diff
@@ -133,17 +165,23 @@ def __handle_deleted_pod_event(self, event: PodEvent) -> None:
         pod_name = pod.metadata.name
         task_metadata = self.task_metadata[pod_name]
 
+        raw_event = event['raw_object'] if event else ''  # TODO: when is this empty?
```
Whoops, forgot to remove this TODO but I did have a question -- do you think it's worth crafting an event string here that indicates this came from reconciliation? 
I could not find where we use this in tron on the kubernetes side, but on the mesos side we strictly used it as a string in error reporting.

nemacysts replied:
`raw_object` is normally just a V1Pod in `dict` form at the moment for `task` events and probably not very useful for most people :p 

I think that if we want to report on reconciliation events, we'd maybe want to send a `control` events, but probably only if we want to take action on the fact that we reconciled/didn't reconcile something

imo, maybe we just want to send `None` in `raw` for reconciled events? that way, the data structures are a little more clear - Optional[Dict] instead of Union[Dict, str]

jfongatyelp replied:
Cool, I have just left these `None` for reconciled events.

nemacysts commented on task_processing/plugins/kubernetes/kubernetes_pod_executor.py:
```diff
@@ -394,7 +432,24 @@ def reconcile(self, task_config: KubernetesTaskConfig) -> None:
+        if not pod:
+            # Pod has gone away while restarting
+            logger.info(f"Pod {pod_name} for task {task_config.name} was no longer found.")
+            self.__update_deleted_pod(pod_name=pod_name, event=None)
```
I think we'll want to add a helper for this case and send a `lost` event instead of treating this like a deleted pod event

nemacysts commented on tests/unit/plugins/kubernetes/kubernetes_pod_executor_test.py:
```diff
@@ -266,3 +281,75 @@ def test_process_event_enqueues_task_processing_events_deleted(
+
+
+def test_reconcile_running_pod(
+    k8s_executor_with_tasks
```
thoughts on having the fixture yield a tuple of (executor, task_configs)?

jfongatyelp replied:
This worked nicely, and it let me test all 4 pod phases we handle much easier.

jfongatyelp commented on task_processing/plugins/kubernetes/kube_client.py:
```diff
@@ -153,3 +153,33 @@ def create_pod(
+                    namespace=namespace, field_selector=f'metadata.name={pod_name}'
+                ).items
+                if not pod:
+                    # XXX: We do not currently distinguish between not finding a pod and hitting api
+                    #      exception when fetching pod
```
Yeah. I added a log line when we explicitly dont' find it so at least there's that, but otherwise downstream code won't know the difference -- is that a big issue?

nemacysts replied:
ah - it's probably worth re-raising the ApiException instead of always returning None

jfongatyelp replied:
Alright I've gone this route though I wasn't clear what we wanted to do on the reconcile() side -- should we try to just remove the pod from task_metadata or leave it be at that point in `TASK_UNKNOWN` forever? 

Will we be trying to reconcile all known running pods on some interval?

jfongatyelp commented on task_processing/plugins/kubernetes/kubernetes_pod_executor.py:
```diff
@@ -45,12 +46,37 @@
+                    self.task_metadata = self.task_metadata.set(
+                        pod_name,
+                        KubernetesTaskMetadata(
+                            task_config=KubernetesTaskConfig(image='', command='noop command'),
```
Yeah this felt weird doing this, I like being able to tell this was explicitly due to reconciling and not some other issue setting image/command.

jfongatyelp commented:
Lots of updates just applied --
* Additional exception handling; we now throw exceptions when we cannot find a pod via k8s API instead of just treating it like a pod that cannot be found
* Tests updated as suggested in review comments
* Fixed multiple instances of not reassigning pmap variables to results of set() (set() does not modify variables in place)
* Now setting KubernetesTaskMetadata.node_name from pod.spec.node_name instead of pod.status.host_ip
* Fixed `get_pod` implementation (leftovers from a previous use of list_namespaced_pods instead of read_namespaced_pod)

jfongatyelp commented:
New updates per our discussion of what we're going to restore from state -- no longer just pod names, but instead taking in full task_configs.

nemacysts commented on task_processing/plugins/kubernetes/kubernetes_pod_executor.py:
```diff
@@ -391,10 +430,53 @@ def run(self, task_config: KubernetesTaskConfig) -> Optional[str]:
+        try:
+            pod = self.kube_client.get_pod(namespace=self.namespace, pod_name=pod_name)
+        except Exception:
+            logger.exception(f"Cannot reconcile pod {pod_name}")
```
should we set `pod = None` here so that we correctly enter line 452?

jfongatyelp replied:
This was originally just `return` ing here but lost the line during some updates, which would have left the action in a perpetual unknown state.  

I think it seems more user-friendly to treat this as LOST, so I'll go your `pod = None` route.

nemacysts commented on task_processing/plugins/kubernetes/kubernetes_pod_executor.py:
```diff
@@ -391,10 +430,53 @@ def run(self, task_config: KubernetesTaskConfig) -> Optional[str]:
+                )
+            )
+
+        with self.task_metadata_lock:
+            if not pod:
+                # Pod has gone away while restarting
+                logger.info(
+                    f"Pod {pod_name} for task {task_config.name} was no longer found. "
+                    "Marking as LOST"
+                )
+                self.task_metadata = self.task_metadata.set(
+                    pod_name,
+                    task_metadata.set(
+                        task_state=KubernetesTaskState.TASK_LOST,
+                        task_state_history=task_metadata.task_state_history.append(
+                            (KubernetesTaskState.TASK_LOST, time.time()),
+                        )
+                    )
+                )
+                self.event_queue.put(
+                    task_event(
+                        task_id=pod_name,
+                        terminal=False,
+                        timestamp=time.time(),
+                        raw=None,
+                        task_config=task_metadata.task_config,
+                        platform_type="lost"
+                    )
+                )
+            else:
+                # Treat like a modified pod
+                self.__update_modified_pod(pod=pod, event=None)
```
just curious: how come we're not just nesting this in the above context manager?

jfongatyelp replied:
Ah, this was an error, there is no reason not to all work within the same context manager.

nemacysts commented on tests/unit/plugins/kubernetes/kubernetes_pod_executor_test.py:
```diff
@@ -266,3 +301,73 @@ def test_process_event_enqueues_task_processing_events_deleted(
+
+    mock_pods = []
+    test_phases = ['Running', 'Succeeded', 'Failed', 'Unknown']
+    for i in range(len(mock_task_configs)):
+        taskconf = mock_task_configs[i]
+        phase = test_phases[i]
```
thoughts on something like
```suggestion
    for taskconf, phase in zip(mock_task_configs, test_phases):
```
explicitly indexing in python always looks a bit off to me, but feel free to drop this issue :)

nemacysts commented on tests/unit/plugins/kubernetes/kubernetes_pod_executor_test.py:
```diff
@@ -266,3 +301,73 @@ def test_process_event_enqueues_task_processing_events_deleted(
+            k8s_executor.reconcile(taskconf)
+
+    assert k8s_executor.event_queue.qsize() == 4
+    assert len(k8s_executor.task_metadata) == 2
```
might be worth adding a comment here about what two tasks we're expecting to be in task_metadata just in case this ever fails in the future since this test has a decent bit of setup and its not immediately clear

nemacysts commented on task_processing/plugins/kubernetes/kubernetes_pod_executor.py:
```diff
@@ -199,7 +238,7 @@ def __handle_modified_pod_event(self, event: PodEvent) -> None:
                 pod_name,
                 task_metadata.set(
-                    node_name=pod.status.host_ip,
+                    node_name=pod.spec.node_name,
```
nice, thank you for fixing this!
